homeassistant/components/gardena_bluetooth/coordinator.py

```diff
@@ -0,0 +1,219 @@
+"""Provides the switchbot DataUpdateCoordinator."""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Awaitable, Callable
+from contextlib import asynccontextmanager
+from datetime import timedelta
+import logging
+from typing import TYPE_CHECKING, Any
+
+from bleak import BleakClient
+from bleak.exc import BleakError
+from bleak_retry_connector import establish_connection
+from gardena_bluetooth import read_char_raw, write_char
+from gardena_bluetooth.exceptions import CharacteristicNoAccess
+from gardena_bluetooth.parse import Characteristic, CharacteristicType
+
+from homeassistant.components import bluetooth
+from homeassistant.core import CALLBACK_TYPE, HassJob, HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.event import async_call_later
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import DOMAIN
+
+if TYPE_CHECKING:
+    pass
+
+SCAN_INTERVAL = timedelta(seconds=60)
+DISCONNECT_DELAY = timedelta(seconds=5)
+LOGGER = logging.getLogger(__name__)
+
+
+class DeviceUnavailable(UpdateFailed, HomeAssistantError):
+    """Raised if device can't be found."""
+
+
+class CallLaterJob:
+    """Helper to contain a function that is to be called later."""
+
+    def __init__(self, hass: HomeAssistant, fun: Callable[[], Awaitable[None]]) -> None:
+        """Initialize a call later job."""
+        self._fun = fun
+        self._hass = hass
+        self._cancel: CALLBACK_TYPE | None = None
+
+        async def _call(_):
+            self._cancel = None
+            await self.call_now()
+
+        self._job = HassJob(_call)
+
+    def cancel(self):
+        """Cancel any pending delay call."""
+        if self._cancel:
+            self._cancel()
+            self._cancel = None
+
+    async def call_now(self):
+        """Call function now."""
+        self.cancel()
+        await self._fun()
+
+    def call_later(self, delay: float | timedelta):
+        """Call function sometime later."""
+        self.cancel()
+        self._cancel = async_call_later(self._hass, delay, self._job)
+
+
+class CachedClient:
```

Comment:
happy to take a PR for bleak-retry-connector for that.  I think a lot of this code can be moved into a lib and out of HA

Reply:
Yup, kind of kept it here since i felt it would make sense to generalize into HA. But i can place it in lib instead, only "problem" is how to do delay scheduled things in lib, but i suppose i can use asyncio standard methods instead of hassjob stuff.

Reply:
Moved into lib. delay calls of async functions is rather ugly in asyncio, but worked out alright. Once i've iterated it a few times, and seen if i can use it for Fjäråskupan too, i'll try to get a pull for the retry-connector lib.